co-log/tutorials/Main.hs

```diff
@@ -34,6 +36,13 @@ example = do
     log D "First message..."
     logInfo "Second message..."
 
+simpleExample :: MonadIO m => RichMsg m SimpleMsg
+simpleExample = RichMessagex (SimpleMsg callStack "Simple text message") defaultFieldMap
+
+simpleApp :: (MonadIO m, WithLog env Text m) => m ()
+simpleApp =
+    logMsg =<< fmtSimpleRichMessageDefault simpleExample
```

Comment:
I think that the example needs to be changed. Let me try to elaborate on how it should look like. 

The main idea behind `SimpleMsg` is that users of logging can provide only textual message and they will automatically see nice things like timestamp. thread id, line code. So instead of old:

```haskell
putStrLn "Some message..."
```

They can use:

```haskell
logMsg "Some message..."
```

And get a much nicer output. This should be possible in theory, because `CallStack` is acquired automatically from the context, and all IO actions are stored in the `FieldMap`. So let's try to achieve that.

First, the example we want to run should log only text message. It should be something like that:

```haskell
simpleMsgExample :: WithLog env SimpleMsg m => m ()
simpleMsgExample = do
    logText "First simple message"
    logText "Second simple message"
```

To run this example, we need to construct `LogAction IO Text` (as you did in `runSimpleApp`). But this action should be constructed differently, so it gets everything automatically. Basically, you need to write a function:

```
toSimpleMsg :: LogAction Text Simpl -> LogAction IO SimpleMsg
```

Which takes constructed `LogAction IO Text` (in our case `textAction` variable) and redirects output of the formatted `SimpleMsg` to `LogAction IO Text`. You also need to implement `logText` function manually which will construct `SimpleMsg` from `Text` by adding `CallStack` to it automatically.

Reply:
@chshersh Thank you a lot for your mentoring and patience. I'm working on it. There's one small piece of the puzzle I'm still missing. The formatter takes a `RichMsg m SimpleMsg` but here with `logText` we're building just a `SimpleMsg`. Is the formatter signature wrong? Should `logText` return a `RichMsg m SimpleMsg` instead of a simple `SimpleMsg`?

Reply:
@sphaso It's not a problem that formatter takes `RichMsg`. The idea here is that conversion to `RichMsg` happens implicitly and transparent for the user. You have:

```haskell
textAction :: LogAction IO Text
```

Then you are going to use [`cmapM`](https://hackage.haskell.org/package/co-log-core-0.2.0.0/docs/Colog-Core-Action.html#v:cmapM) function to create:

```haskell
richAction :: LogAction IO (RichMsg IO SimpleMsg)
richAction = cmapM fmtSimpleRichMessageDefault textAction
```

And then you can use `cmap` and `upgradeMessageAction` to create final action:

```haskell
simpleAction :: LogAction IO SimpleMsg
simpleAction = upgradeMessageAction defaultFieldMap richAction
```

> **NOTE:** For the final part you also need to generalise the type of the `upgradeRichMessage` function

And that's it. So the value flow is the following here:

1. You pass `Text` to `logText` function.
2. `logText` creates `SimpleMsg` from the `Text` and `CallStack` and outputs `SimpleMsg` to the underlying `LogAction`.
3. Than `RichMsg` is automatically created from `SimpleMsg` by attaching `defaultFieldMap` to it.
4. Then you format `RichMsg` as `Text`.
5. And then you output `Text` to the given `textAction` finally.

Reply:
Thank you so much! I think I've finally made it. I've pushed a new commit where I follow the guideline you provided. Let me know if you'd like anything to be changed.